packages/web/src/services/FileDownloadServiceWeb.ts

```diff
@@ -39,6 +45,73 @@ export default class FileDownloadServiceWeb extends HttpServiceBase implements F
         }
     }
 
+    private async downloadS3Directory(fileInfo: FileInfo): Promise<DownloadResult> {
+        const { bucket, key, region } = this.parseS3Url(fileInfo.path);
+        const keys = await this.listS3Objects(bucket, key, region);
+
+        if (keys.length === 0) {
+            throw new Error("No files found in the specified S3 directory.");
+        }
+
+        const fileStream = StreamSaver.createWriteStream(`${fileInfo.name}.zip`);
+        const writer = fileStream.getWriter();
+
+        try {
+            for (const fileKey of keys) {
+                const fileUrl = `https://${bucket}.s3.${region}.amazonaws.com/${encodeURIComponent(
+                    fileKey
+                )}`;
+                const response = await axios.get(fileUrl, { responseType: "blob" });
+
+                const blob = response.data;
+                const stream = blob.stream();
+                await stream.pipeTo(
+                    new WritableStream({
+                        write(chunk) {
+                            writer.write(chunk);
+                        },
+                    })
+                );
+            }
+            writer.close();
+
+            return {
+                downloadRequestId: fileInfo.id,
+                resolution: DownloadResolution.SUCCESS,
+            };
+        } catch (err) {
+            console.error(`Failed to download directory: ${err}`);
+            writer.abort();
+            throw err;
+        }
+    }
+
+    private parseS3Url(url: string): { bucket: string; key: string; region: string } {
+        const { hostname, pathname } = new URL(url);
+        const [bucket] = hostname.split(".");
+        const key = pathname.slice(1);
+        const region = hostname.split(".")[2];
+        return { bucket, key, region };
+    }
+
+    private async listS3Objects(bucket: string, prefix: string, region: string): Promise<string[]> {
```

Comment:
`parseS3Url` & `listS3Objects` seem like they should be put into the super class for this. You can make the super an abstract class so you can add the methods to it.

Reply:
ooh cool didn't think of that. will do.

Reply:
resolved in [cfdf9b4](https://github.com/AllenInstitute/biofile-finder/pull/204/commits/cfdf9b40628c86077ac4ad73f158267bde3b5b49)

Reply:
Left another comment checking in about the way this was abstracted